File: c++/075.cpp
class Solution {
/*
[扫描一趟](https://leetcode.cn/problems/sort-colors/comments/#:~:text=2022%2D06%2D02-,%E5%8F%91%E7%8E%B0%E8%80%83%E7%A0%94%E7%AE%97%E9%A2%98%E7%9A%84%E6%97%B6%E5%80%99%E5%81%9A%E8%BF%87%E7%9A%84%E3%80%82,-1.%E8%AE%BE%E7%BD%AE%E4%B8%89)

设置三个指针：l, r, t;
保证：l 左边（不包括 l）全是0，r 右边（不包括 r）全是2
过程：t 为当前处理的元素
若 nums[t]==0，与 [l] 交换，此时 t++,l++，（交换后一定有 nums[l]==0）
若 nums[t]==1，跳过
若 nums[t]==2，与 [r] 交换，此时 r--, t不变。（因为交换后 nums[t] 是未知的）
*/
public:
    void sortColors(vector<int>& nums) {
        int l=0, r=nums.size()-1, t=0;
        while(t<=r){
            if(nums[t]==0){
                swap(nums[l],nums[t]);
                l++;
                t++;
            }else if(nums[t]==2){
                swap(nums[r],nums[t]);
                r--;
            }else t++;
        }
    }
};

File: go/075.go
package main

/**
颜色分类
红 白 蓝，使用整数 0、 1 和 2 分别表示红色、白色和蓝色。
经典双指针问题
思路：双指针维护三个区间即可
重点在于如何利用双指针去维护
定义 0000 |l| 1111 |r| 2222 即两个边界线
循环不变式，维护指针的意义
l r 代表左右两个分界线 i 循环迭代标记
为了方便迭代 l=-1  r=len  max(i) = r-1
*/
func sortColors(nums []int) {
	l, r := -1, len(nums)
	for i := 0; i < r; {
		if nums[i] == 0 {
			l++
			swap(nums, l, i)
			i++
			continue
		}
		if nums[i] == 1 {
			i++
			continue
		}
		if nums[i] == 2 {
			r--
			swap(nums, r, i)
		}
	}
}
